Sources/Fullscreen/FullscreenGallery/FullscreenGalleryViewController.swift

```diff
@@ -0,0 +1,508 @@
+
+    public override var prefersStatusBarHidden: Bool {
+        return true
+    }
+
+    public weak var galleryDataSource: FullscreenGalleryViewControllerDataSource?
+    public weak var galleryDelegate: FullscreenGalleryViewControllerDelegate?
+
+    // MARK: - Private properties
+
+    private static let captionFadeDuration = 0.2
+    private static let dismissButtonSize: CGFloat = 30.0
+
+    private let viewModel: FullscreenGalleryViewModel
+    private let previewViewInitiallyVisible: Bool
+
+    private var currentImageIndex: Int
+    private var previewViewVisible: Bool
+    private var previewViewWasVisibleBeforePanning = false
+    private var hasPerformedInitialPreviewScroll = false
+
+    private var galleryTransitioningController: FullscreenGalleryTransitioningController? {
+        return transitioningDelegate as? FullscreenGalleryTransitioningController
+    }
+
+    // MARK: - UI properties
+
+    private lazy var backgroundView: UIView = {
+        let view = UIView(frame: .zero)
+        view.backgroundColor = .black
+        view.isUserInteractionEnabled = false
+        return view
+    }()
+
+    private lazy var captionLabel: Label = {
+        let label = Label(style: .title4)
+        label.translatesAutoresizingMaskIntoConstraints = false
+        label.numberOfLines = 0
+        label.textColor = .milk
+        label.textAlignment = .center
+        label.backgroundColor = UIColor.black.withAlphaComponent(0.4)
+        label.shadowOffset = CGSize(width: 1.0, height: 1.0)
+        label.shadowColor = .black
+        return label
+    }()
+
+    private lazy var dismissButton: UIButton = {
+        let button = UIButton(type: .custom)
+        button.translatesAutoresizingMaskIntoConstraints = false
+        button.addTarget(self, action: #selector(dismissButtonTapped), for: .touchUpInside)
+        button.backgroundColor = .clear
+
+        let removeImage = UIImage(named: .remove)
+        button.setImage(removeImage, for: .normal)
+        button.contentHorizontalAlignment = .fill
+        button.contentVerticalAlignment = .fill
+        button.imageView?.contentMode = .scaleAspectFit
+        return button
+    }()
+
+    private lazy var previewView: GalleryPreviewView = {
+        let previewView = GalleryPreviewView()
+        previewView.translatesAutoresizingMaskIntoConstraints = false
+        previewView.delegate = self
+        previewView.dataSource = self
+        return previewView
+    }()
+
+    private lazy var previewViewVisibleConstraint: NSLayoutConstraint = {
+        return previewView.bottomAnchor.constraint(equalTo: view.safeLayoutGuide.bottomAnchor)
+    }()
+
+    private lazy var previewViewHiddenConstraint: NSLayoutConstraint = {
+        // The constant exists to prevent the preview-view from jumping back into the visible area
+        // during the dismissal animation.
+        return previewView.topAnchor.constraint(equalTo: view.bottomAnchor, constant: .mediumLargeSpacing)
+    }()
```

Comment:
Could rather be:

```swift
// The constant exists to prevent the preview-view from jumping back into the visible area
// during the dismissal animation.
private lazy var previewViewHiddenConstraint = previewView.topAnchor.constraint(equalTo: view.bottomAnchor, constant: .mediumLargeSpacing)
```